include/rdma/fabric.h

```diff
@@ -392,6 +392,7 @@ struct fi_tx_attr {
 	size_t			iov_limit;
 	size_t			rma_iov_limit;
 	uint32_t		tclass;
+	uint64_t		optional_caps;
```

Comment:
To support FI_AV_USER_ID, optional_caps was added. This was selected over using the existing caps field due to easier adoption by providers. Specifically, providers should not have to be updated to conform to the defined behavior of optional_caps.

Reply:
I was envisioning using the standard caps field, but categorizing some of the bits as optional caps, similar to what's done for primary/secondary/modifier caps.  Based on this, FI_AV_USER_ID would be defined as a cap bit, but as optional.

This is going a different route, where the app could conceivably list any cap bit as optional.  This is more useful but also more involved.  In this case, the caps bits stay as mandatory if set.  This option could basically remove the need to define any cap bit as 'secondary', since the app can indicate support by setting it as an optional_bit.

We should gather feedback on these options.  The latter (your change) may be appealing to apps, and I don't think ends up that difficult to support.

Reply:
I'm not overly concerned about needing to update the providers.  For the providers that use the utility code for getinfo processing, that code would handle the check(s).  I'm more concerned that if we do break the API that we do that in the best way possible.

Updating the standard caps doesn't break the API.  Plus.  But this series already breaks the API, so we're needing to deal with that anyway.  Providers that don't support FI_AV_USER_ID would need to filter that bit off.  So, I agree that I think this ends up changing all providers to do something with that bit.  Negative.

Adding new optional_caps fields gives us more flexibility and makes things easier on the apps.  Plus.  Most providers shouldn't need updating, since they would end up with optional_caps set to 0.  Plus.  But until a provider does update, it won't include optional_caps as being supported.  Neutral.

My current preference is:

- Define FI_AV_USER_ID as a primary caps bit.
- Add optional_caps as new fields, as you've done.
- Allow requesting any cap bit as optional.
- Apps should only set a given bit in either the caps or optional_caps fields.  If the same bit is set in both, it is treated as mandatory (caps bit takes preference).
- Provider support for caps bits is unchanged.

Handling for optional_caps would be done as:
```
if (user->caps & ~(prov->caps))
    return -FI_ENODATA; // User requires a cap we don't support.

// Get list of optional caps that provider can support
opt_caps = user->optional_caps & prov->caps;

// output info
info->caps = user->caps & prov->caps;
info->caps |= opt_caps;
```